turbonomic/kubeturbo
# [TRB-46371] Support kubeturbo SLO scaling via ORM v2

# Intent
This PR implements support in `kubeturbo` to allow updating workload controller replicas via ORM v2.

# Background
Our existing [ORM v2 improvement](https://jsw.ibm.com/browse/TRB-18317) only implemented support for container resize with kubeturbo. This PR extends the ORM v2 support in `kubeturbo` for horizontal scale.

# Implementation
* Added `actionType` to the `parentController` struct to indicate the type of an action (resize or horizontal scale) associate with a controller.
* Added a predefined map `actionTypeToPathTemplate` to map from controller type and action type to the corresponding resource path. This makes it easy to look up the owned resource path during ORM v2 mapping.
* The core change is in the function `getOwnedResourcePaths`.
* The rest of the changes in this PR are mainly code refactoring.

# Testing
* Make sure full build passes
* Deploy a dev build in Maximo's environment, and execute a horizontal scale action via ORM v2. Confirmed the action is successful, and the following message is logged by `kubeturbo`:
```
I0823 19:59:45.297637       1 remote_mediation_client.go:216] [handleServerMessages][ServerRequestEndpoint] : received message with request type Action.
I0823 19:59:45.297768       1 action_handler.go:368] Received an action HORIZONTAL_SCALE for entity WORKLOAD_CONTROLLER [masinst1-tenant1-mea] in namespace [mas-masinst1-manage]
I0823 19:59:45.297986       1 k8s_controller_updater.go:156] Begin to update Deployment mas-masinst1-manage/masinst1-tenant1-mea ...
I0823 19:59:45.314505       1 k8s_controller_updater.go:191] Try to update replicas of Deployment mas-masinst1-manage/masinst1-tenant1-mea from 1 to 3.
I0823 19:59:45.315325       1 k8s_controller.go:186] Updating Deployment mas-masinst1-manage/masinst1-tenant1-mea via operator for HORIZONTAL_SCALE action ...
I0823 19:59:45.315385       1 k8s_controller.go:194] Detected owned resource paths .spec.replicas for action HORIZONTAL_SCALE.
I0823 19:59:45.315423       1 orm_client_manager.go:122] Found owner resource paths using ORM v2 for owned object Deployment:mas-masinst1-manage:masinst1-tenant1-mea
I0823 19:59:45.315449       1 k8s_controller.go:200] Detected top level owner resources.
I0823 19:59:45.337096       1 orm_client_manager.go:209] Updating owner resource ManageWorkspace/masinst1-tenant1 for Deployment/masinst1-tenant1-mea in namespace mas-masinst1-manage at owner path .spec.settings.deployment.serverBundles[?(@.name=="mea")].replica ...
I0823 19:59:45.684932       1 orm_client_manager.go:221] Successfully updated owner CR ManageWorkspace/masinst1-tenant1 for path '.spec.settings.deployment.serverBundles[?(@.name=="mea")].replica' from 1 to 3 for Deployment/masinst1-tenant1-mea in namespace mas-masinst1-manage.
I0823 19:59:45.684966       1 k8s_controller_updater.go:174] Successfully updated Deployment mas-masinst1-manage/masinst1-tenant1-mea.
I0823 19:59:45.684982       1 horizontal_scaler.go:63] Action HorizontalScale for workload controller[mas-masinst1-manage/masinst1-tenant1-mea] succeeded.
```

# Checklist

These are the items that must be done by the developer and by reviewers before the change is ready to merge. Please ~~strikeout~~ any items that are not applicable, but don't delete them

- [x] Developer Checks
    - [x] Full build with unit tests and fmt and vet checks
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [x] Manual testing done (and described)
    - [ ] Product sweep run and passed
    - [ ] Developer wiki updated (and linked to this description)
- [ ] Reviewer Checks
    - [ ] Merge request description clear and understandable
    - [ ] Developer checklist items complete
    - [ ] Functional code review (how is the code written)
    - [ ] Architectural review (does the code try to do the right thing, in the right way)
    - [ ] Defensive coding (incoming data checked / sanitized, exceptions logged, clear error messages)
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Security review checklist complete.

# Audience

_(@ mention any `review/...` groups or people that should be aware of this merge request)_
